Markdown support in pinned items - improvements

kamilmielnik opened the pull request:
Closes #31816

### Description

This PR addresses post-merge feedback for #31251 which was a fix for #30381

Changes:
- Show pinned item description tooltip only when ellipsis is applied
- Change `<hr>` color in markdown: [screenshot](https://github.com/metabase/metabase/assets/6830683/f012f746-49ad-432e-ba17-b96c129b11ff)
- Change `<hr>` color in markdown tooltip: [screenshot](https://github.com/metabase/metabase/assets/6830683/27f2d680-429f-41f3-8e4e-017060f00588)
  - this affects other markdown tooltips, e.g. these shown when "i" icon is hovered

### Demo

#### Before

https://github.com/metabase/metabase/assets/6830683/19c69f99-529a-413f-964b-713b9cc756ff


#### After

https://github.com/metabase/metabase/assets/6830683/6ca04d40-58c1-4e60-9a72-87a0db5b79fb

kamilmielnik committed:
Extract getIsTruncated

kamilmielnik committed:
Extract useIsTruncated

kamilmielnik committed:
Extract useIsTruncated to a separate file

kamilmielnik committed:
Show tooltip in MarkdownPreview only if the text is truncated

kamilmielnik committed:
Update hr colors in markdown and in markdown tooltip

kamilmielnik committed:
Add a test to show tooltip only when ellipsis is applied

ranquild commented on frontend/src/metabase/core/components/MarkdownPreview/MarkdownPreview.tsx:
```diff
@@ -17,24 +18,38 @@ export const MarkdownPreview = ({
+     * react-markdown API does not allow passing ref to the container div.
+     * We can acquire the reference through its parent.
+     */
+    const reactMarkdownRoot = divRef?.querySelector("div");
```
btw, does it always a `div` or should we get the first element?

kamilmielnik replied:
Always a div: https://github.com/remarkjs/react-markdown/blob/main/lib/react-markdown.js#L120

kamilmielnik replied:
Yeah, in our case it's always a div because we always have `className`: https://github.com/remarkjs/react-markdown/blob/main/lib/react-markdown.js#L120

kamilmielnik replied:
Updated to check for first element for extra safety: c9ca931eec164aae549667cc127a494226253c9a

ranquild commented on frontend/src/metabase/core/components/MarkdownPreview/MarkdownPreview.tsx:
```diff
@@ -17,24 +18,38 @@ export const MarkdownPreview = ({
+}: Props) => {
+  const { isTruncated, ref } = useIsTruncated();
+
+  const setReactMarkdownRef: LegacyRef<HTMLDivElement> = divRef => {
```
maybe `setReactMarkdownRef = (div: HTMLElement | null) => ref.current = div ?div.firstElementChild : null;`

ranquild commented on frontend/src/metabase/hooks/use-is-truncated.ts:
```diff
@@ -0,0 +1,36 @@
+
+import resizeObserver from "metabase/lib/resize-observer";
+
+export const useIsTruncated = <E extends HTMLElement>() => {
```
Can it be `Element` and not `HTMLElement`?

kamilmielnik replied:
If we make it `Element`, then we cannot use `ofsetWidth`, we'd have to go with `clientWidth`. 

There is an inconsistency in `getIsTruncated` (I didn't change it in this PR):
- for height we're checking `clientHeight`
- for width we're checking `offsetWidth` (not `clientWidth`)

I think this may have been an intentional decision (to account for the scrollbar) but I found no info about it in commit history.

Nonetheless, I think we should be using `clientWidth`. This change will affect all places `Ellipsified` is used in, though. 
And then we can use `Element` here. 

Good to go?

kamilmielnik replied:
72e61b5c99b5315c104940ddf434cc9b2f409701

ranquild commented on frontend/src/metabase/core/components/MarkdownPreview/MarkdownPreview.styled.tsx:
```diff
@@ -10,3 +12,10 @@ export const TruncatedMarkdown = styled(Markdown)`
 `;
+
+export const TooltipMarkdown = styled(Markdown)`
+  hr {
```
I think we should pass something to `Markdown` component to make it style this differently. To consolidate all markdown styles in one place

kamilmielnik replied:
Nice catch :+1: There were other places that needed this.

 adea9aa15da4adf33fa874ca0fca366714ec152d

ranquild reviewed (approved):
Looks good, minor comments

kamilmielnik committed:
Use Element instead of HTMLElement in resize-observer and useIsTruncated
- Use clientWidth instead of offsetWidth to detect horizontal ellipsis (offsetWidth comes from HTMLElement.prototype)

kamilmielnik committed:
Introduce "dark" prop to Markdown and make tooltips use is consistently throughout the code

kamilmielnik committed:
Use first child instead element of a div query selector to hopefully improve code safety

kamilmielnik committed:
Update failing tests with ellipsis emulation
